### PR TITLE
feat(cvsb-18931): update date validation and recent expiry logic - changes on branch 18928

### PR DESCRIPTION
**Pull request by m-ujaffer**
Changes to handle expiry date format for legacy data and to throw error if any other format of date occurs.

Original ticket - (changes on 18928 branch still)
https://jira.dvsacloud.uk/browse/CVSB-18928 
closed in favour of
https://jira.dvsacloud.uk/browse/CVSB-18931

### CODE REVIEW

**Review by JonLazariniDD — COMMENTED** [review]
It is the wrong title in the PR (18931) i think or wrong branch name (18928)?

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 1]
```diff
@@ -22,25 +23,31 @@ describe("TestDataProvider", () => {
...
-          expect.assertions(1);
           expect(error).toEqual(
-            new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR)
+            new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR)
           );
         }
       });
     });
-    describe("when getBySystemNumber returns invalid testEndTimestamp", () => {
+    describe("when getBySystemNumber returns invalid testStartTimestamp", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:32]: The testEndTimestamp is invalid in this test

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 2]
```diff
@@ -49,17 +56,87 @@ describe("TestDataProvider", () => {
...
         } catch (error) {
-          expect.assertions(1);
           expect(error).toEqual(
-            new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR)
+            new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR)
           );
         }
       });
     });
+    describe("when getBySystemNumber returns invalid testEndTimestamp", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:65]: Starttime is invalid :)

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 3]
```diff
@@ -49,17 +56,87 @@ describe("TestDataProvider", () => {
...
+            return {
+              getBySystemNumber: () => {
+                return {
+                  Count: 2,
+                  Items: [
+                    {
+                      testStartTimestamp: "2019-03-10T08:47:59.269Z",
+                      testEndTimestamp: "2019-03-10T09:30:59.269Z",
+                      testTypes: [
+                        { testCode: "fft1", testExpiryDate: "2019-02-10T08:47:59.261009Z" }
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:113]: Can we flip this to being the same short format as the start/end.  Then we have one entry looking like a legacy test and the other like an app one.  Covers both basis

**Review by bristol-mike-jones — CHANGES_REQUESTED** [review]
I think its only the always casting errors to a 500 that needs actioning the others are improvements / personal prefs.

File: src/handlers/expiry/providers/DateProvider.ts [hunk 4]
```diff
@@ -24,14 +24,22 @@ export class DateProvider {
...
   public static getEpoc(): Date {
-  return moment(new Date(0)).startOf("day").toDate();
+    return moment(new Date(0)).startOf("day").toDate();
   }
+
   public static isSameAsEpoc(inputDate: Date): boolean {
     return moment(inputDate).isSame(DateProvider.getEpoc());
   }
 
+  public static getInstance(dateValue: string | number | Date) {
```
Comment (JonLazariniDD) [inline src/handlers/expiry/providers/DateProvider.ts:36]: can it receive a number or Date as input? I can that getInstance is called after the date is validated with isValidDate and this function checks for 3 patterns. How would the isValidate function behaves if it receives other formats than the ones we explicitly check?
I wonder whether we should be adding unit tests for this function which will act as some documentation for other developers

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 5]
```diff
@@ -22,61 +17,66 @@ export class TestDataProvider implements ITestDataProvider {
...
           );
       }
       return result;
     } catch (error) {
-      if (!(error instanceof HTTPError)) {
-        console.log(error);
-        error = new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR);
-      }
+      console.error("getTestHistory: Error ", error);
+      error = new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR);
```
[the remark this answers is not in the document]
Comment (JonLazariniDD) [reply src/handlers/expiry/providers/TestDataProvider.ts:46]: getTestHistory is called by getMostRecentExpiryDate which is called by the controller (generateExpiryDate()) which is already catching the error and throwing 500 if it fails (or the DAO fails too) with async, I don't think it will be necessary to check on HTTPError however do we want to be more meaningful about the error we return to differentiate between a service failure or failure to calculate the date because of the data we have received?

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 6]
```diff
@@ -22,61 +17,66 @@ export class TestDataProvider implements ITestDataProvider {
...
+              }
+            console.log(`getMostRecentExpiryDate: Filtered Date -> ${testExpiryDate}`);
+            filteredTestTypeDates.push(DateProvider.getInstance((testExpiryDate)));
           }
-        });
+
       });
-      if (filteredTestTypeDates.length) {
-        maxDate = moment.max(filteredTestTypeDates).toDate();
+        if (filteredTestTypeDates.length) {
```
Comment (JonLazariniDD) [inline src/handlers/expiry/providers/TestDataProvider.ts:69]: I think we need to add prettier at some point

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 7]
```diff
@@ -22,25 +23,31 @@ describe("TestDataProvider", () => {
...
           expect(error).toEqual(
-            new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR)
+            new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR)
           );
         }
       });
     });
-    describe("when getBySystemNumber returns invalid testEndTimestamp", () => {
+    describe("when getBySystemNumber returns invalid testStartTimestamp", () => {
       it("should throw error", async () => {
```
Comment (JonLazariniDD) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:33]: could we add a bit more information when it should throw the error in the description please ?
Is that when `testEndTimestamp ` is incorrectly formatted?

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 8]
```diff
@@ -49,17 +56,87 @@ describe("TestDataProvider", () => {
...
-          expect.assertions(1);
           expect(error).toEqual(
-            new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR)
+            new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR)
           );
         }
       });
     });
+    describe("when getBySystemNumber returns invalid testEndTimestamp", () => {
+      it("should throw error", async () => {
```
Comment (JonLazariniDD) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:66]: same here, is that when `testStartTimestamp ` has an invalid format?

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 9]
```diff
@@ -49,17 +56,87 @@ describe("TestDataProvider", () => {
...
+          const result = await testDataProvider.getTestHistory("123");
+          console.log(result);
+        } catch (error) {
+          expect(error).toEqual(
+            new HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR)
+          );
+        }
+      });
+    });
+    describe("when getBySystemNumber returns array of valid data", () => {
```
Comment (JonLazariniDD) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:99]: do we want to add one more when we have 000000 too so we cover our 3 patterns?

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts:46]: revert the change.

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 6 above)
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts:69]: Will be handled in CVSB-18734

File: tests/unit/expiry/TestDataProvider.unitTest.ts (hunk 9 above)
Comment (m-ujaffer) [reply tests/unit/expiry/TestDataProvider.unitTest.ts:99]: rephrase the description with format and add scenario with 3 digits.

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 10]
```diff
@@ -149,7 +225,7 @@ describe("TestDataProvider", () => {
     });
 
     describe("when getBySystemNumber returns an array of valid and invalid recent expiry dates", () => {
```
Comment (m-ujaffer) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:227]: Add a test for null/undefined expiry dates.

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 11]
```diff
@@ -22,61 +17,66 @@ export class TestDataProvider implements ITestDataProvider {
...
-              testCode.toUpperCase()
-            ) &&
-            testExpiryDate &&
-            DateProvider.isValidDate(testExpiryDate)
-          ) {
-            filteredTestTypeDates.push(moment(testExpiryDate));
+
+          if (testCode && TestDataProvider.isValidTestCodeForExpiryCalculation(testCode.toUpperCase()) && testExpiryDate) {
+            if (!DateProvider.isValidDate(testExpiryDate)) {
+                throw new Error(`Invalid Expiry Date: ${testExpiryDate}`);
```
Comment (alex-deloitte) [inline src/handlers/expiry/providers/TestDataProvider.ts:62]: As discussed earlier, can we please check if this is behaviour is in line with the business expectations?

Comment (bristol-mike-jones) [reply src/handlers/expiry/providers/TestDataProvider.ts:62]: I've double-checked this with the PO and he doesn't want to change this behaviour.  Good catch @alex-deloitte 👍 

I've sent an email to this effect and asked for the ticket to be updated.

**Review by alex-deloitte — APPROVED** [review]

File: tests/unit/expiry/TestDataProvider.unitTest.ts (hunk 1 above)
Comment (m-ujaffer) [reply tests/unit/expiry/TestDataProvider.unitTest.ts]: fixed

File: tests/unit/expiry/TestDataProvider.unitTest.ts (hunk 2 above)
Comment (m-ujaffer) [reply tests/unit/expiry/TestDataProvider.unitTest.ts]: fixed

File: tests/unit/expiry/TestDataProvider.unitTest.ts (hunk 3 above)
Comment (m-ujaffer) [reply tests/unit/expiry/TestDataProvider.unitTest.ts]: changed to match the app format.

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 11 above)
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts]: reverted the change and the tests. Only logs are maintained.

**Review by bristol-mike-jones — APPROVED** [review]

**Review by bristol-mike-jones — APPROVED** [review]

**Review by JonLazariniDD — APPROVED** [review]
nice thanks

changes on branch 18928 however ticket is closed (see ticket 18931 for requirements)
